torch/csrc/jit/tensorexpr/ir_simplifier.h

```diff
@@ -63,304 +309,732 @@ static const Expr* newBinaryOpOfType(
-    if (const Name##Imm* imm = dynamic_cast<const Name##Imm*>(expr)) { \
-      val = imm->value();                                              \
-    } else {                                                           \
-      LOG(FATAL) << "Bad expr: " << *expr << "\n";                     \
-    }                                                                  \
-    break;
-    AT_FORALL_SCALAR_TYPES_AND2(Half, Bool, TYPE_CASE);
-#undef TYPE_CASE
-    default:
-      LOG(FATAL) << "Unsupported datatype: " << expr->dtype();
+      const Expr* ret =
+          new Broadcast(new Op(bc->value(), bcother->value()), bc->lanes());
+      return ret->accept_mutator(mutator);
+    }
+
+    if (const Ramp* r = dynamic_cast<const Ramp*>(rhs)) {
+      if (bc->lanes() != r->lanes()) {
+        throw malformed_input("multilane lane mismatch");
+      }
+
+      const Expr* ret =
+          new Ramp(new Op(bc->value(), r->base()), r->stride(), r->lanes());
+      return ret->accept_mutator(mutator);
+    }
+  } else if (const Ramp* ramp = dynamic_cast<const Ramp*>(lhs)) {
+    if (const Ramp* rother = dynamic_cast<const Ramp*>(rhs)) {
+      if (ramp->lanes() != rother->lanes()) {
+        throw malformed_input("multilane lane mismatch");
+      }
+
+      const Expr* ret = new Ramp(
+          new Op(ramp->base(), rother->base()),
+          new Op(ramp->stride(), rother->stride()),
+          ramp->lanes());
+      return ret->accept_mutator(mutator);
+    }
+
+    if (const Broadcast* bc = dynamic_cast<const Broadcast*>(rhs)) {
+      if (ramp->lanes() != bc->lanes()) {
+        throw malformed_input("multilane lane mismatch");
+      }
+      const Expr* ret = new Ramp(
+          new Op(bc->value(), ramp->base()), ramp->stride(), ramp->lanes());
+      return ret->accept_mutator(mutator);
+    }
   }
 
-  return val;
+  return nullptr;
 }
 
-/* Takes a LinearForm and converts it to Mul + (Add/Sub). */
-const Expr* expandLinearForm(const LinearForm* v, IRMutator* mutator) {
-  const Expr* mul = nullptr;
-  const Expr* A = v->getA();
-  const Expr* B = v->getB();
-  const Expr* X = v->getX();
-  // we only really care about 0 and 1, so double should be fine.
-  double Aval = immediateAs<double>(A);
-  double Bval = immediateAs<double>(B);
-
-  // First handle A.
-  if (Aval == 0) {
-    if (Bval == 0) {
-      return getImmediateByType(X->dtype(), 0).node();
-    }
-    return B;
-  } else if (Aval == 1) {
-    mul = X;
-  } else if (Aval == -1) {
-    return new Sub(B, X);
-  } else if (Aval < 0) {
-    // Negate A.
-    ExprHandle zero = getImmediateByType(A->dtype(), 0);
-    Sub* A_Sub = new Sub(zero.node(), A);
-
-    return new Sub(B, new Mul(X, evaluateOp(A_Sub)));
-  } else {
-    mul = new Mul(X, A);
-  }
-
-  if (Bval == 0) {
-    return mul;
-  }
-
-  return new Add(mul, B);
+// Handles optimization cases for Broadcast/Ramp * Broadcast/Ramp
+const Expr* mulMultilane(const Expr* lhs, const Expr* rhs, IRMutator* mutator) {
```

Comment:
`ret->accept_mutator(mutator)` can be moved out of this function to places where it's called to keep the signature of this function cleaner.